envpool/mujoco/dmc/mujoco_env.h

```diff
@@ -0,0 +1,196 @@
+  }
+
+  // https://github.com/deepmind/dm_control/blob/1.0.2/dm_control/rl/control.py#L94
+  void ControlStep(const mjtNum* action) {
+    TaskBeforeStep(action);
+    PhysicsStep(n_sub_steps_);
+    TaskAfterStep();
+    reward_ = TaskGetReward();
+    if (++elapsed_step_ >= max_episode_steps_) {
+      discount_ = 1.0;
+      done_ = true;
+    } else if (TaskShouldTerminateEpisode()) {
+      // copy from composer task
+      // this is different because we don't use None in C++
+      discount_ = 0.0;
+      done_ = true;
+    } else {
+      discount_ = TaskGetDiscount();
+      done_ = false;
+    }
+  }
+
+  // Task
+  void TaskInitializeEpisodeMjcf() {}
+  void TaskInitializeEpisode() {}
+  // https://github.com/deepmind/dm_control/blob/1.0.2/dm_control/suite/base.py#L73
+  void TaskBeforeStep(const mjtNum* action) { PhysicsSetControl(action); }
+  void TaskBeforeSubStep(const mjtNum* action) {}
+  void TaskAfterStep() {}
+  void TaskAftersSubStep() {}
+  float TaskGetReward() {
+    throw std::runtime_error("GetReward not implemented");
+  }
+  float TaskGetDiscount() { return 1.0; }
+  bool TaskShouldTerminateEpisode() { return false; }
+
+  // Physics
+  // https://github.com/deepmind/dm_control/blob/1.0.2/dm_control/mujoco/engine.py#L263
+  void PhysicsReset(int keyframe_id = -1) {
+    if (keyframe_id < 0) {
+      mj_resetData(model_.get(), data_.get());
+    } else {
+      // actually no one steps to this line
+      assert(keyframe_id < model_->nkey);
+      mj_resetDataKeyframe(model_.get(), data_.get(), keyframe_id);
+    }
+
+    // PhysicsAfterReset may be overwritten?
+    int old_flags = model_->opt.disableflags;
+    model_->opt.disableflags |= mjDSBL_ACTUATION;
+    PhysicsForward();
+    model_->opt.disableflags = old_flags;
+  }
+
+  // https://github.com/deepmind/dm_control/blob/1.0.2/dm_control/mujoco/engine.py#L286
+  void PhysicsAfterReset() {
+    int old_flags = model_->opt.disableflags;
+    model_->opt.disableflags |= mjDSBL_ACTUATION;
+    PhysicsForward();
+    model_->opt.disableflags = old_flags;
+  }
+
+  // https://github.com/deepmind/dm_control/blob/1.0.2/dm_control/mujoco/engine.py#L138
+  void PhysicsSetControl(const mjtNum* control) {
+    std::memcpy(data_->ctrl, control, sizeof(mjtNum) * model_->nu);
+  }
+
+  // https://github.com/deepmind/dm_control/blob/1.0.2/dm_control/mujoco/engine.py#L292
+  void PhysicsForward() { mj_forward(model_.get(), data_.get()); }
+
+  // https://github.com/deepmind/dm_control/blob/1.0.2/dm_control/mujoco/engine.py#L146
+  void PhysicsStep(int nstep) {
+    if (model_->opt.integrator != mjINT_RK4) {
+      mj_step2(model_.get(), data_.get());
+      if (nstep > 1) {
+        for (int i = 0; i < nstep - 1; ++i) {
+          mj_step(model_.get(), data_.get());
+        }
+      }
+    } else {
+      for (int i = 0; i < nstep; ++i) {
+        mj_step(model_.get(), data_.get());
+      }
+    }
+    mj_step1(model_.get(), data_.get());
+  }
+};
+
+}  // namespace mujoco
```

Comment:
Just add a base class according to the discussion about #103
cc @Benjamin-eecs @kevinzakka

Reply:
Not sure when TaskBeforeSubStep and TaskAftersSubStep are called in suite env.